### PR TITLE
steam_support: generalize pressure-vessel root paths

### PR DESCRIPTION
**Pull request by ashuntu**
This PR changes every `pressure-vessel` parent to `*` (previously just `run`).

This fixes https://github.com/canonical/steam-snap/issues/356 in the Steam snap, where users were encountering the error: `bwrap: Can't mount tmpfs on /newroot/var/pressure-vessel/ldso: Permission denied` when running any Proton game, which represents a huge portion of games.

I included every pressure-vessel path (not just the tmpfs one noted in the issue) since pressure-vessel is likely to change at any point, as noted by @smcv here: https://github.com/canonical/steam-snap/issues/356#issuecomment-1893784766. In that comment, Simon recommended something more general and less restrictive (`/newroot/**/`), but for the time being I've opted for a more restrictive approach here.

Security permitting, we could generalize pressure-vessel paths more, e.g. `/newroot/*/pressure-vessel/**,` or even `/newroot/**,`. I'm not sure of the full security implications of these options, so I would love to have a discussion on how this can be improved.

*For reference, [Pressure Vessel](https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/tree/main/pressure-vessel) is a containerization tool used by Steam to run games inside. Steam and Pressure Vessel are generally to be considered trusted.*

### CODE REVIEW

File: interfaces/builtin/steam_support.go
Comment (smcv) [inline interfaces/builtin/steam_support.go:142]: At the risk of repeating myself, these paths are an implementation detail. It is very likely that we will find that we need to do something slightly different in a future release.

For instance, if we find that we also need to bind-mount `ld.so.conf.d` because otherwise the container runtime framework doesn't work on, I don't know, Gentoo or something, then we will have to start doing that; and if snapd doesn't like that, then Steam will just not work.

Comment (kenvandine) [reply interfaces/builtin/steam_support.go:142]: I think @smcv is right, we've seen paths like this change it would be great to make them as general as possible.  I'd prefer `/newroot/**,`

Comment (ashuntu) [reply interfaces/builtin/steam_support.go:142]: Could we rely on the `pressure-vessel` directory, so, something like `/newroot/*/pressure-vessel/**,`? If `/newroot/**,` is the solution here, was there a reason that the initial rules were so restrictive (maybe @kenvandine knows that context, I'm not sure)?

Comment (kenvandine) [reply interfaces/builtin/steam_support.go:142]: We usually lean more towards restrictive rules, so specifying more of the path is a common pattern.  However, steam-support is more permissive than most interfaces, so it would be good to accept the nature of that, IMO.  We'll have to get a security review anyway, so I'd recommend starting there and we can tighten it down more if necessary.

Comment (ashuntu) [reply interfaces/builtin/steam_support.go:142]: I've changed each pressure-vessel based path to `/newroot/**,`

Comment (smcv) [reply interfaces/builtin/steam_support.go:142]: > I've changed each pressure-vessel based path to `/newroot/**,`

Thanks, what you have done for e.g. the Wayland and X11 sockets looks like a big improvement.

It seems a little ironic that as far as I can see, you haven't done this for the line I started this thread from? :-)

File: interfaces/builtin/steam_support.go
Comment (smcv) [inline interfaces/builtin/steam_support.go:190]: This is redundant with the new ` -> /newroot/**` below?

File: interfaces/builtin/steam_support.go
Comment (ashuntu) [reply interfaces/builtin/steam_support.go]: > It seems a little ironic that as far as I can see, you haven't done this for the line I started this thread from? :-)

Silly mistake, fixed now :)

File: interfaces/builtin/steam_support.go
Comment (ashuntu) [reply interfaces/builtin/steam_support.go:202]: Yeah, I purposely didn't widen this any further for that reason. We could widen it if it's ok security-wise

Comment (smcv) [reply interfaces/builtin/steam_support.go:202]: I think probably the restriction you ideally want here is something like "after doing the `pivot_root` into its new root directory, pressure-vessel can write wherever it wants to write" - but I suspect that isn't actually representable, because AppArmor only sees paths relative to the current root directory, and doesn't know when we did the `pivot_root`.

Assuming that pressure-vessel will use a directory named `pressure-vessel` (somewhere) for its internal state doesn't seem too bad in terms of probability of regressing. For a long time we were using `/run/pressure-vessel`, but we had to stop using that path for the `ld.so.cache`, for better compatibility with an x86-on-ARM CPU emulation framework (FEX-Emu) that fakes a chroot/overlayfs in user-space.

**Comment by smcv** [comment]
While reviews are necessary for `steam_support`, taking the opportunity to resolve https://github.com/canonical/steam-snap/issues/359 might be a good idea.

**Comment by smcv** [comment]
Some more things to look at in this AppArmor profile:

`/sys/module/nvidia/version r` would also be very useful, for example for https://github.com/canonical/steam-snap/issues/361.

`/etc/debian_chroot r` is a nice-to-have, for https://github.com/canonical/steam-snap/issues/360.

**Comment by ashuntu** [comment]
I'll definitely look at implementing those other changes for the `steam_support` profile and Steam in other PRs. I think we should probably keep this PR scoped as-is just to get the pressure-vessel apparmor fix in as soon as possible.

**Comment by smcv** [comment]
> I think we should probably keep this PR scoped as-is just to get the pressure-vessel apparmor fix in as soon as possible

Sure, I just wanted to make sure the other AppArmor stuff was on your (and reviewers') radar.

File: interfaces/builtin/steam_support.go
Comment (ashuntu) [reply interfaces/builtin/steam_support.go:200]: I opened an issue in the Steam snap to track this: [PR_SET_NO_NEW_PRIVS](https://github.com/canonical/steam-snap/issues/362)

**Comment by alexmurray** [comment]
> is there a way that Steam will be able to discover, programmatically, that it is running under a suitable snapd version?

The steam snap could specify the required snapd version via `assumes` - https://snapcraft.io/docs/snapcraft-top-level-metadata#heading--assumes

**Review by kenvandine — APPROVED** [review]
This should do wonders for future proofing the steam-support interface to future pressure vessel changes, thanks!

**Comment by kenvandine** [comment]
@alexmurray this is still waiting for a security review, could you please take a look at this?

**Comment by smcv** [comment]
> I am a little concerned about changing the mount rules to just target `/newroot/**`: it may solve the immediate problems, but we may just end up repeating them if PV changes the source paths for mounts next.

Well, yes, Canonical has chosen to package Steam as a Snap app, so Canonical gets to bear the maintenance cost of that app; and the design of Snap, combined with the design of Steam, means that every time there's a change in implementation details, a reaction to that change is quite likely to be needed.

Part of the problem here is that we're trying to put pressure-vessel's square peg into Snap's round hole: Snap wants to control what paths an app can and can't access, and that's always going to be hard to reconcile with using a mount namespace to rewrite what the filesystem means. https://github.com/canonical/steam-snap/issues/365 has some thoughts about a possible alternative to this approach.

My reasoning for suggesting `/newroot/**` as the target is that if Snap would be willing to give Steam read/write access to any particular path in the absence of nested containers, then there's no security benefit to demanding that the path is mounted at any specific location inside the nested container - the worst-case scenario from a security point of view would be that Steam chooses a path inside the nested container that will give it read-write access, but we've already assumed that read-write access is acceptable, so that's fine. So the only thing you're going to get from trying to limit the target is increasing Snap's exposure to regressing as a result of Steam implementation changes.

I didn't suggest `/oldroot/** ->` even though there definitely *are* more paths that Steam is going to want to access (https://github.com/canonical/steam-snap/issues/359) because if I imagine that I was a security reviewer, then an unrestricted source path is something that I would want to be more cautious about. But, instead of being hyper-specific about the paths that Steam literally does access right now, I'd prefer the designers of this interface to be thinking in terms of: which paths would or wouldn't it be OK if Steam hypothetically *did* access in future? - and then preemptively giving at least read-only access to relatively large sections of filesystem that wouldn't be considered to be a problem.

**Comment by jslarraz** [comment]
Hi,

To the best of my understanding and after looking at the PR, it does not seem to enlarge the attack surface significantly compared with the currently deployed version. After discussion with @alexmurray we agreed that it is fine to approve the PR security wise. 

Thanks

**Review by jslarraz — APPROVED** [review]

**Review by alexmurray — APPROVED** [review]
LGTM!

**Review by jhenstridge — APPROVED** [review]
If Alex is happy with this, then it looks good.

If we want to relax the mount rules further for PV further without granting more permission to other processes, we might want to consider a sub-profile. entered when executing `pv-wrap`, and return to the main profile when `pv-wrap` executes the game. The subprofile could then be given all the permissions needed to manipulate mount namesapces without granting it to every other process in the snap.

This would still be somewhat advisory, since Steam downloads PV from its store, and a malicious game could potentially create an executable that matches whatever AppArmor rules we write for the `pv-wrap` subprofile transition.

**Review by ernestl — APPROVED** [review]
Thanks

**Review by pedronis — APPROVED** [review]
no objections

**Pull request merged**
